File: Medium-Array/StockandBuy.java
// WAP to buy and sell stocks 
import java.util.Scanner;
public class StockandBuy
{
    public static int stockBuySell(int arr[]) 
    {
        // for single transtation 
        // int min=Integer.MAX_VALUE;
        // int profit=0;
        // for(int i=0;i<arr.length;i++)
        // {
        //     if(arr[i]<min)
        //     {
        //         min=arr[i];
        //     }
        //     else
        //     {
        //         profit=Math.max(profit,arr[i]-min);
        //     }
        // }
        // return profit;
        // for multiple transaction
        int profit = 0;
        for (int i = 1; i < arr.length; i++)
        {
            if (arr[i] > arr[i - 1]) 
            {
                profit += arr[i] - arr[i - 1]; // Buy at i-1, sell at i
            }
        }
        return profit;
    }
    public static void main(String[] args)
    {
       // Create a Scanner object to take input from the user
       Scanner sc=new Scanner(System.in);
       System.out.println("Size of Array");
       // Read the size of the array
       int n=sc.nextInt();
       int arr[]=new int[n];
       System.out.println("Enter the elements of array");
       // Read the elements of the array
       for(int i=0;i<n;i++)
       {
        arr[i]=sc.nextInt();
       }
       System.out.println("Maximum Profit is "+stockBuySell(arr));
    }
}

File: Medium-Array/StockandBuy.py
# WAP to to buy and sell stock to maximize profit. You can only complete at most one transaction (i.e., buy one and sell one share of the stock).

# function to calculate the maximum profit from a list of stock prices
def maxProfit(arr):
    # check if the list of prices is empty
    if not arr:
        return 0
    
    # initialize the minimum price to the first price in the list and the maximum profit to 0
    min_price = arr[0]
    max_profit = 0

    # iterate through the list of prices
    for price in arr:
        if price < min_price:
            min_price = price
        else:
            max_profit = max(max_profit, price - min_price)

    # return the maximum profit
    return max_profit

arr=[]
n=int(input("Enter the number of stock prices: "))
print("Enter the stock prices: ")
for i in range(n):
    price=int(input())
    arr.append(price)
profit = maxProfit(arr)
print("Maximum profit: ", profit)